### PR TITLE
DT-463 Call community-api to fake update custody record in Delius

### PR DESCRIPTION
**Pull request by andymarke**
- added call to actually update Delius (though that service is also still a dummy service that just logs what it would do rather than actually doing it)
- additional checks on movements associated with bookings that are no longer current (we see a handful of these being done by mistake)

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt [hunk 1]
```diff
@@ -12,10 +16,26 @@ open class CommunityService(@Qualifier("communityApiRestTemplate") private val r
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
-  open fun updateProbationCustody(offender: Offender) {
+  open fun updateProbationCustody(offenderNo: String, bookingNo: String, updateCustody: UpdateCustody): Custody? {
+    try {
+      val response = restTemplate.exchange("/secure/offenders/nomsNumber/{nomsNumber}/custody/bookingNumber/{bookingNumber}", HttpMethod.PUT, HttpEntity(updateCustody), Custody::class.java, offenderNo, bookingNo)
+      return response.body!!
+    } catch (e : HttpClientErrorException) {
+      if (e.statusCode != HttpStatus.NOT_FOUND) throw e
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt]: 404 is valid response, it means NOMIS record doesn't link with Delius record

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt [hunk 2]
```diff
@@ -20,10 +22,21 @@ open class OffenderService(@Qualifier("elite2ApiRestTemplate") val restTemplate:
...
         return response.body!!
     }
+
+    open fun getMovement(bookingId: Long, movementSeq: Long): Movement? {
+        try {
+            val response = restTemplate.getForEntity("/api/bookings/{bookingId}/movement/{movementSeq}", Movement::class.java, bookingId, movementSeq)
+            return response.body!!
+        } catch (e : HttpClientErrorException) {
+            if (e.statusCode != HttpStatus.NOT_FOUND) throw e
+            // 404 is "valid" since it means movement is for an inactive booking
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt]: 404 is valid response (well actually EliteAPI currently returns 404 when booking is no longer active which is possibly not 100% correct)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt [hunk 3]
```diff
@@ -46,8 +59,13 @@ data class Prisoner(
 data class Movement(
         val offenderNo: String,
         val createDateTime: LocalDateTime,
-        val fromAgency: String?,
-        val toAgency: String?,
+        val fromAgency: String,
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt]: these will never be null

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementService.kt [hunk 4]
```diff
@@ -22,32 +22,69 @@ open class PrisonMovementService(private val offenderService: OffenderService, p
...
-                            "toAgency" to movement.toAgency,
-                            "offenderNo" to prisoner.offenderNo,
-                            "firstName" to prisoner.firstName,
-                            "lastName" to prisoner.lastName,
-                            "latestLocation" to prisoner.latestLocation,
-                            "convictedStatus" to prisoner.convictedStatus
-                    ),
-                    null)
+        if (movement != null) {
+            if (isMovementTransferIntoPrison(movement)) {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementService.kt]: This is magic; because of the `if (movement != null)` the Kotlin compiler now knows all references to movement below is non-null so we can write `movement.fromAgency` rather than `movement?.fromAgency` - amazing

**Comment by andymarke** [comment]
Service still does nothing other then log events on what it will do after we have finished analysing that what is will do is correct.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt [hunk 5]
```diff
@@ -12,10 +16,26 @@ open class CommunityService(@Qualifier("communityApiRestTemplate") private val r
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
-  open fun updateProbationCustody(offender: Offender) {
+  open fun updateProbationCustody(offenderNo: String, bookingNo: String, updateCustody: UpdateCustody): Custody? {
+    try {
```
Comment (brightonsbox) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt]: minor comment, spacing here (2 spaces instead of 4) seems inconsistent with the rest of the codebase

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt [hunk 6]
```diff
@@ -20,10 +22,21 @@ open class OffenderService(@Qualifier("elite2ApiRestTemplate") val restTemplate:
...
 
-    open fun getMovement(bookingId: Long, movementSeq: Long): Movement {
-        val response = restTemplate.getForEntity("/api/bookings/{bookingId}/movement/{movementSeq}", Movement::class.java, bookingId, movementSeq)
+    open fun getBooking(bookingId: Long): Booking {
+        val response = restTemplate.getForEntity("/api/bookings/{bookingId}?basicInfo=true", Booking::class.java, bookingId)
         return response.body!!
     }
+
+    open fun getMovement(bookingId: Long, movementSeq: Long): Movement? {
+        try {
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt]: I think the "Kotlin way" is to put the return in front of the try rather than on the rather than at the returned values.  That said, I think it reads better the way you have done it!  I never much liked the "last statement is the return value" approach.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt (hunk 5 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt]: yep, reformatted

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt (hunk 6 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/OffenderService.kt]: yep, just noticed that when IntelliJ hinted I do precisely that. Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementService.kt [hunk 7]
```diff
@@ -22,32 +22,69 @@ open class PrisonMovementService(private val offenderService: OffenderService, p
...
+                val booking = offenderService.getBooking(bookingId)
+
+                if (booking.activeFlag) {
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementService.kt]: This block of ifs and elses is becoming a bit hard to follow... I'm not sure if this code is for the long term or just to assist with analysis, but if it's going to stick around it might be good to play with the when statement and make it look pretty?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementServiceTest.kt [hunk 8]
```diff
@@ -109,11 +181,11 @@ class PrisonMovementServiceTest {
             imprisonmentStatus = "",
             receptionDate = "")
 
-    private fun createPrisonAdmissionMovement(offenderNo: String) = Movement (
+    private fun createPrisonAdmissionMovement(offenderNo: String = "AB123D", toAgency: String = "MDI") = Movement (
             offenderNo = offenderNo,
             createDateTime = LocalDateTime.now(),
             fromAgency = "ABRYCT",
-            toAgency = "MDI",
+            toAgency = toAgency,
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementServiceTest.kt]: Named parameters are just great.  Who needs builders anymore?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt (hunk 5 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt]: It's probably worth mentioning that Peter previously told me to change my Intellij settings to 2 spaces for Kotlin.  But probably one for another PR...
image

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementService.kt (hunk 7 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/PrisonMovementService.kt]: Yes, I thinking the same thing. Mostly they will go away - but I'm finding it very noisy. Let's pair for a bit and see how we can improve it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt (hunk 5 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/CommunityService.kt]: Thanks, maybe I'll do a PR with just formatting changes
